README.md
# Go Identity Library v2

Golang library for Identity Service.

## FAQ 
- Difference with go-identity-lib (v1)?
Instead of using `tenant` as logical separator, this library uses `namespace`

CHANGELOG.md
# Changelog

## Unreleased
### v0.0.1-alpha.x
- Auth filter that match between requested permission and claimed permission
- Refresh token

filter/auth_filter_test.go
package filter

import (
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"

	"github.com/agriardyan/identity-lib-v2/claim"
	"github.com/dgrijalva/jwt-go"
	"github.com/emicklei/go-restful"

	"github.com/stretchr/testify/assert"
)

func TestMatchOneClaim(t *testing.T) {
	auth := NewSymmetric("testKey")

	claimedResource := []string{"external", "fnb", "namespaceA", "kiosk", "123456", "customer"}
	requiredResource := []string{"external", "fnb", "namespaceA", "kiosk", "123456", "customer"}
	testResult := auth.matchOneClaim(claimedResource, requiredResource)
	assert.Equal(t, true, testResult)

	claimedResource = []string{"external", "fnb", "namespaceA", "kiosk", "123456", "customer"}
	requiredResource = []string{"external", "fnb", "namespaceA", "kiosk", "*", "customer"}
	testResult = auth.matchOneClaim(claimedResource, requiredResource)
	assert.Equal(t, true, testResult)

	claimedResource = []string{"external", "fnb", "namespaceA", "kiosk", "*", "customer"}
	requiredResource = []string{"external", "fnb", "namespaceA", "kiosk", "12345", "customer"}
	testResult = auth.matchOneClaim(claimedResource, requiredResource)
	assert.Equal(t, true, testResult)

	claimedResource = []string{"external", "fnb", "*", "kiosk", "*", "customer"}
	requiredResource = []string{"external", "fnb", "namespaceA", "kiosk", "12345", "customer"}
	testResult = auth.matchOneClaim(claimedResource, requiredResource)
	assert.Equal(t, true, testResult)

	claimedResource = []string{"external", "fnb", "namespaceA", "kiosk", "*"}
	requiredResource = []string{"external", "fnb", "namespaceA", "kiosk", "12345"}
	testResult = auth.matchOneClaim(claimedResource, requiredResource)
	assert.Equal(t, true, testResult)

	claimedResource = []string{"external", "fnb", "namespaceA", "kiosk", "123456", "customer"}
	requiredResource = []string{"internal", "fnb", "namespaceA", "kiosk", "*", "customer"}
	testResult = auth.matchOneClaim(claimedResource, requiredResource)
	assert.Equal(t, false, testResult)
}

var DummyHandlerResult = ""

func DummyHandler(initialResource string) func(req *restful.Request, resp *restful.Response) {
	return func(req *restful.Request, resp *restful.Response) {
		auth := NewSymmetric("testKey")
		DummyHandlerResult = auth.replacePlaceholders(initialResource, req)
	}
}

func TestReplacePlaceholder(t *testing.T) {

	initialResource1 := "external:fnb:namespace:{namespace}:menu"
	initialResource2 := "external:fnb:namespace:{namespace}:kiosk:{kioskId}"
	initialResource3 := "external:fnb:namespace:{namespace}:kiosk:{kioskId}"
	initialResource4 := "external:fnb:namespace:{namespace}:kiosk:*"

	ws := new(restful.WebService)
	ws.Consumes(restful.MIME_XML)
	ws.Route(ws.GET("/external/{namespace}").To(DummyHandler(initialResource1)))
	ws.Route(ws.GET("/external/{namespace}/kiosk/{kioskId}/menu").To(DummyHandler(initialResource2)))
	ws.Route(ws.GET("/external/{namespace}/kiosk/{kioskId}/menu/people/{peopleId}").To(DummyHandler(initialResource3)))
	ws.Route(ws.GET("/external/{namespace}/kiosk/{kioskId}").To(DummyHandler(initialResource4)))
	restful.Add(ws)

	bodyReader := strings.NewReader("")
	httpRequest, _ := http.NewRequest("GET", "/external/abcnamespace", bodyReader)
	httpRequest.Header.Set("Content-Type", restful.MIME_XML)
	httpWriter := httptest.NewRecorder()

	restful.DefaultContainer.ServeHTTP(httpWriter, httpRequest)

	// this is rather stupid approach for test, but gorestful didn't write the interface and I'm too lazy to wrap restful.Request just for test
	expectedResource := "external:fnb:namespace:abcnamespace:menu"
	assert.Equal(t, expectedResource, DummyHandlerResult, "Should replace placeholder")

	httpRequest, _ = http.NewRequest("GET", "/external/abcnamespace/kiosk/kioskabc/menu", bodyReader)
	restful.DefaultContainer.ServeHTTP(httpWriter, httpRequest)

	expectedResource = "external:fnb:namespace:abcnamespace:kiosk:kioskabc"
	assert.Equal(t, expectedResource, DummyHandlerResult, "Should replace placeholder")

	httpRequest, _ = http.NewRequest("GET", "/external/abcnamespace/kiosk/kioskabc/menu/people/john", bodyReader)
	restful.DefaultContainer.ServeHTTP(httpWriter, httpRequest)

	expectedResource = "external:fnb:namespace:abcnamespace:kiosk:kioskabc"
	assert.Equal(t, expectedResource, DummyHandlerResult, "Should not append anything")

	httpRequest, _ = http.NewRequest("GET", "/external/abcnamespace/kiosk/kioskabc", bodyReader)
	restful.DefaultContainer.ServeHTTP(httpWriter, httpRequest)

	expectedResource = "external:fnb:namespace:abcnamespace:kiosk:*"
	assert.Equal(t, expectedResource, DummyHandlerResult, "Should not change the *")
}

func TestMatchWithClaims(t *testing.T) {
	auth := NewSymmetric("testKey")

	testClaimedPerm := make(map[string]int)
	testClaimedPerm["internal:hrm:namespace:abcnamespace:kiosk:*"] = 1
	testClaimedPerm["internal:hrm:namespace:abcnamespace:menu"] = 3
	testClaimedPerm["external:fnb:namespace:abcnamespace:discount"] = 15

	requiredResource := "internal:hrm:namespace:*:kiosk:*"
	result := auth.matchWithClaims(requiredResource, 1, testClaimedPerm)
	assert.Equal(t, true, result, "Should match")

	requiredResource = "internal:hrm:namespace:*:menu"
	result = auth.matchWithClaims(requiredResource, 1, testClaimedPerm)
	assert.Equal(t, true, result, "Should match")

	requiredResource = "internal:hrm:namespace:*:menu"
	result = auth.matchWithClaims(requiredResource, 2, testClaimedPerm)
	assert.Equal(t, true, result, "Should match")

	requiredResource = "internal:hrm:namespace:*:menu"
	result = auth.matchWithClaims(requiredResource, 8, testClaimedPerm)
	assert.Equal(t, false, result, "Should not match")

	requiredResource = "internal:fnb:namespace:*:discount"
	result = auth.matchWithClaims(requiredResource, 8, testClaimedPerm)
	assert.Equal(t, false, result, "Should not match")
}

type testInput_authorizationHandler struct {
	req                *restful.Request
	requiredPermission string
	requiredAction     int
}

type testExpected_authorizationHandler struct {
	claim *claim.IdentityClaim
	err   error
}

type testScenario_authorizationHandler struct {
	message  string
	auth     *AuthFilter
	input    testInput_authorizationHandler
	expected testExpected_authorizationHandler
}

func TestAuthorizationHandler(t *testing.T) {
	userClaim := &claim.IdentityClaim{
		Permissions: map[string]int{
			"svc:namespace:*:resource":  15,
			"svc:namespace:*:resource2": 2,
			"svc:namespace:*:resource3": 8,
		},
	}
	auth := NewSymmetric("123")
	authWithParser := NewSymmetricWithParser("123", func(tokenString string, keyFunc jwt.Keyfunc) (*jwt.Token, *claim.IdentityClaim, error) {
		claimObj := userClaim
		return nil, claimObj, nil
	})
	authWithExpiredParser := &AuthFilter{
		secretKey: "123",
		jwtParserFunc: func(tokenString string, keyFunc jwt.Keyfunc) (*jwt.Token, *claim.IdentityClaim, error) {
			return nil, nil, jwt.NewValidationError("", jwt.ValidationErrorExpired)
		},
	}
	authWithInvalidToken := &AuthFilter{
		secretKey: "123",
		jwtParserFunc: func(tokenString string, keyFunc jwt.Keyfunc) (*jwt.Token, *claim.IdentityClaim, error) {
			return nil, nil, jwt.NewValidationError("", jwt.ValidationErrorSignatureInvalid)
		},
	}
	testScenarios := []testScenario_authorizationHandler{
		{
			message: "Request header does not contains Authorization",
			auth:    auth,
			input: testInput_authorizationHandler{
				req:                restful.NewRequest(&http.Request{Header: http.Header{"Authorization": []string{""}}}),
				requiredPermission: "svc:namespace:*:resource",
				requiredAction:     2,
			},
			expected: testExpected_authorizationHandler{
				err: ErrUnauthorized,
			},
		},
		{
			message: "Authorization header does not contains Bearer token",
			auth:    auth,
			input: testInput_authorizationHandler{
				req:                restful.NewRequest(&http.Request{Header: http.Header{"Authorization": []string{"Basic ..."}}}),
				requiredPermission: "svc:namespace:*:resource",
				requiredAction:     2,
			},
			expected: testExpected_authorizationHandler{
				err: ErrUnauthorized,
			},
		},
		{
			message: "Token is expired",
			auth:    authWithExpiredParser,
			input: testInput_authorizationHandler{
				req:                restful.NewRequest(&http.Request{Header: http.Header{"Authorization": []string{"Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyLCJleHAiOjE2MjY0OTk2MDB9.arPjT1st5PHKxjPyPnr6u9qruseA2ph6PUWNlIrMes4"}}}),
				requiredPermission: "svc:namespace:*:resource",
				requiredAction:     2,
			},
			expected: testExpected_authorizationHandler{
				err: ErrExpiredToken,
			},
		},
		{
			message: "Token is invalid",
			auth:    authWithInvalidToken,
			input: testInput_authorizationHandler{
				req:                restful.NewRequest(&http.Request{Header: http.Header{"Authorization": []string{"Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyLCJleHAiOjE2MjY0OTk2MDB9.arPjT1st5PHKxjPyPnr6u9qruseA2ph6PUWNlIrMes4"}}}),
				requiredPermission: "svc:namespace:*:resource",
				requiredAction:     2,
			},
			expected: testExpected_authorizationHandler{
				err: ErrInvalidToken,
			},
		},
		{
			message: "Delete is forbidden because not enough privilege",
			auth:    authWithParser,
			input: testInput_authorizationHandler{
				req:                restful.NewRequest(&http.Request{Header: http.Header{"Authorization": []string{"Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyLCJleHAiOjE2MjY0OTk2MDAsInBlcm1pc3Npb25zIjp7InN2YzpuYW1lc3BhY2U6KjpyZXNvdXJjZSI6Miwic3ZjOm5hbWVzcGFjZToqOnJlc291cmNlMiI6MTV9fQ.qr3aUPEU0ESXpRyDE3Bjbe2CD6dSBIgm2jokz4vzukw"}}}),
				requiredPermission: "svc:namespace:*:resource2",
				requiredAction:     8,
			},
			expected: testExpected_authorizationHandler{
				err: ErrForbidden,
			},
		},
		{
			message: "Read is allowed",
			auth:    authWithParser,
			input: testInput_authorizationHandler{
				req:                restful.NewRequest(&http.Request{Header: http.Header{"Authorization": []string{"Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyLCJleHAiOjE2MjY0OTk2MDAsInBlcm1pc3Npb25zIjp7InN2YzpuYW1lc3BhY2U6KjpyZXNvdXJjZSI6Miwic3ZjOm5hbWVzcGFjZToqOnJlc291cmNlMiI6MTV9fQ.qr3aUPEU0ESXpRyDE3Bjbe2CD6dSBIgm2jokz4vzukw"}}}),
				requiredPermission: "svc:namespace:*:resource",
				requiredAction:     2,
			},
			expected: testExpected_authorizationHandler{
				claim: userClaim,
				err:   nil,
			},
		},
		{
			message: "Delete is allowed",
			auth:    authWithParser,
			input: testInput_authorizationHandler{
				req:                restful.NewRequest(&http.Request{Header: http.Header{"Authorization": []string{"Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyLCJleHAiOjE2MjY0OTk2MDAsInBlcm1pc3Npb25zIjp7InN2YzpuYW1lc3BhY2U6KjpyZXNvdXJjZSI6Miwic3ZjOm5hbWVzcGFjZToqOnJlc291cmNlMiI6MTV9fQ.qr3aUPEU0ESXpRyDE3Bjbe2CD6dSBIgm2jokz4vzukw"}}}),
				requiredPermission: "svc:namespace:*:resource3",
				requiredAction:     8,
			},
			expected: testExpected_authorizationHandler{
				claim: userClaim,
				err:   nil,
			},
		},
	}

	for _, testScenario := range testScenarios {
		claim, err := testScenario.auth.authorizeHandler(testScenario.input.req, testScenario.input.requiredPermission, testScenario.input.requiredAction)
		assert.Equal(t, testScenario.expected.claim, claim)
		assert.Equal(t, testScenario.expected.err, err)
	}
}


filter/auth_filter.go
package filter

import (
	"encoding/base64"
	"errors"
	"net/http"
	"strings"
	"time"
	"unicode"

	"github.com/agriardyan/identity-lib-v2/claim"
	"github.com/agriardyan/identity-lib-v2/token"
	"github.com/dgrijalva/jwt-go"
	"github.com/emicklei/go-restful"
	"github.com/sirupsen/logrus"
)

var ErrUnauthorized = errors.New("unauthorized")
var ErrInvalidToken = errors.New("invalidtoken")
var ErrExpiredToken = errors.New("expiredtoken")
var ErrForbidden = errors.New("forbidden")

const (
	HeaderParameterAuthorization      = "Authorization"
	Admin                             = "admin"
	BasicTokenType                    = "Basic"
	BearerTokenType                   = "Bearer"
	UserID                            = "userId"
	ClientID                          = "clientId"
	ClientSecret                      = "clientSecret"
	UnauthenticatedRequestCode        = 1002001
	UnauthenticatedRequestExplanation = "Unauthenticated request"
	InvalidTokenCode                  = 1002002
	InvalidTokenExplanation           = "Invalid token"
	ExpiredTokenCode                  = 1002003
	ExpiredTokenExplanation           = "Expired token"
	ForbiddenRequestCode              = 1003001
	ForbiddenRequestExplanation       = "Forbidden request. Check your privilege!"
	NamespacePlaceholder              = "{namespace}"
	UserPlaceholder                   = "{userId}"
	ResourceSeparator                 = ":"
	StarValue                         = "*"
)

//AuthResponse auth response model
type AuthResponse struct {
	Code        int    `json:"code"`
	Explanation string `json:"explanation"`
}

//IAuthFilter contract for auth filter
type IAuthFilter interface {
	Auth(requiredPermission string, requiredAction int) restful.FilterFunction
	BasicAuth() restful.FilterFunction
}

//AuthFilter auth filter
type AuthFilter struct {
	secretKey     string
	jwtParserFunc func(tokenString string, keyFunc jwt.Keyfunc) (*jwt.Token, *claim.IdentityClaim, error)
}

//NewSymmetric create filter instance with symmetric signing
func NewSymmetric(secretKey string) *AuthFilter {
	return NewSymmetricWithParser(secretKey, token.ParseWithClaims)
}

//NewSymmetricFilter create filter instance with symmetric signing and custom parser
func NewSymmetricWithParser(secretKey string, jwtParserFunc func(tokenString string, keyFunc jwt.Keyfunc) (*jwt.Token, *claim.IdentityClaim, error)) *AuthFilter {
	return &AuthFilter{
		secretKey:     secretKey,
		jwtParserFunc: jwtParserFunc,
	}
}

//Auth authenticate and authorize the request
func (auth *AuthFilter) Auth(requiredPermission string, requiredAction int) restful.FilterFunction {
	return func(req *restful.Request, resp *restful.Response, chain *restful.FilterChain) {

		claimObj, err := auth.authorizeHandler(req, requiredPermission, requiredAction)

		switch err {
		case ErrUnauthorized:
			_ = resp.WriteHeaderAndJson(
				http.StatusUnauthorized,
				&AuthResponse{
					Code:        UnauthenticatedRequestCode,
					Explanation: UnauthenticatedRequestExplanation,
				},
				restful.MIME_JSON,
			)
			return
		case ErrExpiredToken:
			_ = resp.WriteHeaderAndJson(
				http.StatusUnauthorized,
				&AuthResponse{
					Code:        ExpiredTokenCode,
					Explanation: ExpiredTokenExplanation,
				},
				restful.MIME_JSON,
			)
			return
		case ErrInvalidToken:
			_ = resp.WriteHeaderAndJson(
				http.StatusUnauthorized,
				&AuthResponse{
					Code:        ExpiredTokenCode,
					Explanation: ExpiredTokenExplanation,
				},
				restful.MIME_JSON,
			)
			return
		case ErrForbidden:
			_ = resp.WriteHeaderAndJson(
				http.StatusForbidden,
				&AuthResponse{
					Code:        ForbiddenRequestCode,
					Explanation: ForbiddenRequestExplanation,
				},
				restful.MIME_JSON,
			)
			return
		}

		req.SetAttribute(claim.RequesterUserID, claimObj.UserID)

		chain.ProcessFilter(req, resp)
	}
}

func (auth *AuthFilter) authorizeHandler(req *restful.Request, requiredPermission string, requiredAction int) (*claim.IdentityClaim, error) {
	rawToken := req.HeaderParameter(HeaderParameterAuthorization)
	splittedToken := strings.Fields(rawToken)

	if len(splittedToken) != 2 {
		return nil, ErrUnauthorized
	}

	if splittedToken[0] != BearerTokenType {
		return nil, ErrUnauthorized
	}

	jwtKey := []byte(auth.secretKey)

	_, claimObj, err := auth.jwtParserFunc(splittedToken[1], func(token *jwt.Token) (interface{}, error) {
		return jwtKey, nil
	})
	if err != nil {
		if errVal, ok := err.(*jwt.ValidationError); ok {
			if errVal.Errors == jwt.ValidationErrorExpired {
				return nil, ErrExpiredToken
			}
		}
		logrus.Errorf("Failed to parse claim. Error: %s", err)
		return nil, ErrInvalidToken
	}

	requiredPermissionPlaceholderReplaced := auth.replacePlaceholders(requiredPermission, req)
	claimedPerms := claimObj.Permissions
	isAuthorized := auth.matchWithClaims(requiredPermissionPlaceholderReplaced, requiredAction, claimedPerms)
	if !isAuthorized {
		return nil, ErrForbidden
	}

	return claimObj, nil
}

//replacePlaceholders replace the placeholder from expected resource with one from request path param
//for example: assuming the req.PathParameter(namespace) return 12345
//the resource external:fnb:namespace:{namespace}:menu will become external:fnb:namespace:123456:menu
func (auth *AuthFilter) replacePlaceholders(expectedResource string, req *restful.Request) string {
	newResource := ""
	splittedRes := strings.Split(expectedResource, ResourceSeparator)
	for _, res := range splittedRes {
		if strings.HasPrefix(res, "{") && strings.HasSuffix(res, "}") {
			trimmed := strings.TrimFunc(res, func(r rune) bool {
				return !unicode.IsLetter(r)
			})
			pathParam := req.PathParameter(trimmed)
			if pathParam != "" {
				res = pathParam
			}
		}
		newResource += res + ResourceSeparator
	}
	finalTrim := strings.Trim(newResource, ResourceSeparator)
	return finalTrim
}

//matchWithClaims match the permission
//remember that resource format is visibility:servicename:param1:value1:param2:value2
//for example: internal:fnb:namespace:*:user:*:menu or just namespace:{namespace}:menu:*
func (auth *AuthFilter) matchWithClaims(requiredResource string, requiredAction int, claimedPerms map[string]int) bool {

	if claimedPerms[requiredResource]&requiredAction > 0 { // required and granted perfectly match
		return true
	}

	requiredResourceSubs := strings.Split(requiredResource, ResourceSeparator)
	for claimedResource, claimedAction := range claimedPerms {
		claimedResourceSubs := strings.Split(claimedResource, ResourceSeparator)
		if len(claimedResourceSubs) != len(requiredResourceSubs) {
			continue
		}
		if auth.matchOneClaim(claimedResourceSubs, requiredResourceSubs) {
			return requiredAction&claimedAction > 0
		}
	}
	return false
}

func (auth *AuthFilter) matchOneClaim(claimedResourceSubs, requiredResourceSubs []string) bool {
	for i, requiredResourceSub := range requiredResourceSubs {
		if requiredResourceSub != claimedResourceSubs[i] && claimedResourceSubs[i] != StarValue && requiredResourceSub != StarValue {
			return false
		}
	}
	return true
}

func (auth *AuthFilter) isExpire(claims *claim.IdentityClaim) bool {
	now := time.Now().Unix()
	return claims.ExpiresAt <= now
}

//BasicAuth perform parse and validation for basic auth, and put credential to request attribute
//note that you still needs to validate the credential against what you have in DB. This function does not do that for you.
func (auth *AuthFilter) BasicAuth() restful.FilterFunction {
	return func(req *restful.Request, resp *restful.Response, chain *restful.FilterChain) {
		rawToken := req.HeaderParameter(HeaderParameterAuthorization)
		splittedToken := strings.Fields(rawToken)

		responseBodyUnauthenticated := &AuthResponse{
			Code:        UnauthenticatedRequestCode,
			Explanation: UnauthenticatedRequestExplanation,
		}

		if len(splittedToken) != 2 {
			_ = resp.WriteHeaderAndJson(
				http.StatusUnauthorized,
				responseBodyUnauthenticated,
				restful.MIME_JSON,
			)
			return
		}

		if splittedToken[0] != BasicTokenType {
			_ = resp.WriteHeaderAndJson(
				http.StatusUnauthorized,
				responseBodyUnauthenticated,
				restful.MIME_JSON,
			)
			return
		}

		contentBytes, err := base64.StdEncoding.DecodeString(splittedToken[1])
		if err != nil {
			logrus.Errorf("Failed to parse the basic auth. Error: %v", err)
			_ = resp.WriteHeaderAndJson(
				http.StatusUnauthorized,
				&AuthResponse{
					Code:        InvalidTokenCode,
					Explanation: InvalidTokenExplanation,
				},
				restful.MIME_JSON,
			)
			return
		}

		content := string(contentBytes)
		contents := strings.Split(content, ":")

		if len(contents) < 1 || len(contents) > 2 {
			_ = resp.WriteHeaderAndJson(
				http.StatusUnauthorized,
				responseBodyUnauthenticated,
				restful.MIME_JSON,
			)
			return
		}

		clientID := contents[0]
		clientSecret := contents[1]
		req.SetAttribute(ClientID, clientID)
		req.SetAttribute(ClientSecret, clientSecret)

		chain.ProcessFilter(req, resp)
	}
}
